addon/PageScraper.js

```diff
@@ -60,6 +64,26 @@ PageScraper.prototype = {
     }
   }),
 
+  _computeImageSize: Task.async(function*(metadata) {
+    this._setUpImageListener();
+    if (metadata.images[0]) {
+      Services.mm.broadcastAsyncMessage("get-image-size", {url: metadata.images[0].url});
```

Comment:
The `idl` seems to make `broadcastAsyncMessage` quite scary. https://dxr.mozilla.org/mozilla-central/source/dom/base/nsIMessageManager.idl#313-314 I would assume this broadcasts to all frame scripts, so that seems a bit wasteful.

You should be able to just use `new Services.appShell.hiddenDOMWindow.Image();` and use a regular `Promise` instead of the obsolete `defer` form. E.g.,

```js
compute(metadata) {
  return new Promise(resolve => {
    if (images) {
      let image = new Image(); ...
      image.addEventListener("load", () => resolve(image.width));
    }
    else {
      resolve([]);
    }
  });
}
```